packages/java/endpoint/src/main/java/com/vaadin/hilla/signals/core/SignalsRegistry.java

```diff
@@ -0,0 +1,206 @@
+     *
+     * @return the signal instance, or null if no signal is found for the
+     *         provided {@code clientSignalId}
+     * @throws NullPointerException
+     *             if {@code clientSignalId} is null
+     */
+    public synchronized NumberSignal get(UUID clientSignalId) {
+        Objects.requireNonNull(clientSignalId,
+                "Client signal id must not be null");
+        UUID signalId = clientSignalToSignalMapping.get(clientSignalId);
+        if (signalId == null) {
+            LOGGER.debug("No associated signal found for client signal id: {}",
+                    clientSignalId);
+            return null;
+        }
+        return signals.get(signalId);
+    }
+
+    /**
+     * Get a signal instance by the provided {@code signalId}.
+     * <p>
+     *
+     * @param signalId
+     *            the signal id, must not be null
+     *
+     * @return the signal instance, or null if no signal is found for the
+     *         provided {@code signalId}
+     * @throws NullPointerException
+     *             if {@code signalId} is null
+     */
+    public synchronized NumberSignal getBySignalId(UUID signalId) {
+        Objects.requireNonNull(signalId, "Signal id must not be null");
+        return signals.get(signalId);
+    }
+
+    /**
+     * Checks if a mapping exists between a registered signal instance and the
+     * provided {@code clientSignalId}.
+     *
+     * @param clientSignalId
+     *            the client signal id, must not be null
+     * @return true if the signal instance is registered, false otherwise
+     * @throws NullPointerException
+     *             if {@code signalId} is null
+     */
+    public synchronized boolean contains(UUID clientSignalId) {
+        Objects.requireNonNull(clientSignalId,
+                "Client signal id must not be null");
+        if (!clientSignalToSignalMapping.containsKey(clientSignalId)) {
+            return false;
+        }
+        var signalId = clientSignalToSignalMapping.get(clientSignalId);
+        if (!signals.containsKey(signalId)) {
+            throw new IllegalStateException(String.format(
+                    "A mapping for client Signal exists, but the signal itself is not registered. Client signal id: %s",
+                    clientSignalId));
+        }
+        return true;
+    }
+
+    /**
+     * Removes a signal instance by the provided {@code signalId}.
+     * <p>
+     * It also removes all the possible associated client signals, too.
+     *
+     * @param signalId
+     *            the signal id, must not be null
+     * @throws NullPointerException
+     *             if {@code signalId} is null
+     */
+    public synchronized void unregister(UUID signalId) {
+        Objects.requireNonNull(signalId,
+                "Signal id to remove must not be null");
+        signals.remove(signalId);
+        clientSignalToSignalMapping.values().removeIf(signalId::equals);
```

Comment:
This is not a complete review, but I feel that checking all values is a potential performance problem.
I propose this alternative:

- remove `clientSignalToSignalMapping` completely and store all mapping in the existing weak map, no matter if the UUIDs come from the client or the server;
- add a `Set` of related known UUIDs to the `Signal` itself;
- each time a mapping is added, also add the UUID to the corresponding `Set` in the `Signal`;
- when a client UUID is removed from the map, also remove it from the corresponding `Set`;
- when a server UUID is removed from the map, clear the `Signal`'s `Set` completely.

So, all keys related to a `Signal` will be preserved as they are referenced in its `Set`, but the server-side removal still allows full garbage collection, as the client UUIDs will only remain as weak keys.

I think that this implementation is far simpler if made by subclassing `Map`.

Reply:
If you want the server UUID to stay "weak", a second `put` method is also needed, as for removal, to avoid adding that UUID to the `Signal`'s `Set`.

Reply:
Thanks for the review! 
TBH, while I was implementing the SignalsRegistry this way, I wasn't thinking about the performance that much as suggested by the [issue's description](https://github.com/vaadin/hilla/issues/2429). But, I had two things in mind: 
1. To easily find out whether a signal instance returned by an endpoint has already been registered to avoid duplicate registration (the weak map used for this purpose).
2. To easily access the instance to which a client is subscribed (`clientSignalToSignalMapping` is used for this purpose). This access is happening every time a client changes the signal's value, so it needs to be fast.

In a subsequent PR, we will enable clients to unsubscribe from a signal instance, e.g. when the user moves away from the view containing a usage (accessing/manipulating) for that signal instance. This will be supported by the registry's `removeClientSignalToSignalMapping` method (which is fast as well). The `unregister` method would not be called so often, for example, as of now it is not even called anywhere, and it is only there for the sake of feature completeness of the registry itself, so this will not be a performance issue for now. In fact, unregistering a server signal instance can/should happen when there are no client is subscribing to them, which means no mappings in the `removeClientSignalToSignalMapping`, and this will let the garbage collection to remove it from the weak map automatically / eventually.

Your suggested implementation is also supporting both concerns mentioned above, however, I find it easier to debug with the this way, where all the mappings are stored in a single class (maybe it is a matter of taste). Maybe we can even remove the `unregister` method, and address that functionality and its possible performance issues later if happens. WDYT?